### PR TITLE
Add application system

### LINKED ISSUE #24: Add application (+ management) pages
aka move https://osu.spotlights.team to here and stop using trello/discord

### PR DESCRIPTION
**Pull request by venix12**
Resolves #24 

Stuffs that need to be done before:
- [x] restructure auth permissions system #32 
- [x] don't allow more than one app for one gamemode in one app cycle, possibly some usergroups?
- [x] feedback pm templates?
- [x] ~~voting system? (to be discussed...)~~
- [x] question types

### CODE REVIEW

File: app/AppCycle.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    ];
+
+    /**
+     * Scopes
+     */
+
+    public function scopeCurrent($query)
+    {
+        return $query->where('active', true)->first();
+    }
```
Comment (cl8n) [inline app/AppCycle.php]: this isn't a scope
```php
public static function current()
{
    return static::where('active', true)->first();
}
```
it also depends on there always being only one active, which doesn't seem to be guaranteed anywhere (it'd be best to guarantee at model-level)

File: app/AppCycle.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * Methods
+     */
+
+    public static function isActive()
+    {
+        return count(static::where('active', true)->get()) > 0;
```
Comment (cl8n) [inline app/AppCycle.php]:
```suggestion
        return static::where('active', true)->exists();
```

File: app/Http/Controllers/ApplicationController.php [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+        $app = Application::create([
+            'cycle_id' => AppCycle::current()->id,
+            'gamemode' => $fields['gamemode'],
+            'user_id' => auth()->user()->id,
+        ]);
+
+        foreach ($fields as $key => $value)
+        {
+            if ($key !== '_token' && $key !== 'gamemode') {
```
Comment (cl8n) [inline app/Http/Controllers/ApplicationController.php]: should have the answers come in their own array so you don't have to guess which query options are the right ones.

File: app/Http/Controllers/Admin/AppEvaluationController.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        $apps = $cycle->applications;
+
+        $appsCollection = fractal_transform($apps, 'ApplicationTransformer');
+
+        return view('admin.app-eval.show')
+            ->with('appsCollection', $appsCollection)
+            ->with('cycle', $cycle);
+    }
+
+    public function storeCycle()
```
Comment (cl8n) [inline app/Http/Controllers/Admin/AppEvaluationController.php]: i dont really get the point if the new cycle isn't active by default and marks all the others as inactive. you can't make this work without editing db unless i missed somethign?

File: app/Application.php [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+{
+    protected $casts = [
+        'approved' => 'boolean',
+    ];
+
+    protected $fillable = [
+        'approved',
+        'cycle_id',
+        'feedback',
+        'feedback_author_id',
```
Comment (cl8n) [inline app/Application.php]: each application can only receive a single user's feedback? is this temporary?

File: app/Http/Controllers/ApplicationController.php [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+use App\Event;
+
+class ApplicationController extends Controller
+{
+    public function __construct()
+    {
+        $this->middleware('auth');
+    }
+
+    public function index()
```
Comment (cl8n) [inline app/Http/Controllers/ApplicationController.php]: following laravel semantics this is a `create()`

File: database/migrations/2020_02_09_193140_create_app_cycles_table.php [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('app_cycles', function (Blueprint $table) {
+            $table->bigIncrements('id');
```
Comment (cl8n) [inline database/migrations/2020_02_09_193140_create_app_cycles_table.php]: definitely doesnt need big, could be a lot smaller

File: database/migrations/2020_02_09_201925_create_app_answers_table.php [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::create('app_answers', function (Blueprint $table) {
+            $table->bigIncrements('id');
+            $table->bigInteger('app_id')->unsigned();
+            $table->bigInteger('question_id')->unsigned();
+            $table->string('answer', 3000);
```
Comment (cl8n) [inline database/migrations/2020_02_09_201925_create_app_answers_table.php]: the 3000 won't actually do anything in this application

File: routes/web.php [hunk 9]
```diff
@@ -69,6 +84,11 @@
     });
 });
 
+Route::group(['prefix' => 'app'], function () {
+    Route::get('/', 'ApplicationController@index')->name('aop-form');
```
Comment (cl8n) [inline routes/web.php]: name typo

**Comment by cl8n** [comment]
are all of those checklist items going to be in this branch?

**Comment by venix12** [comment]
nope, if anything, pm templates only

File: database/migrations/2020_02_09_201925_create_app_answers_table.php (hunk 8 above)
Comment (venix12) [reply database/migrations/2020_02_09_201925_create_app_answers_table.php]: seems like it will tho

File: app/Application.php (hunk 5 above)
Comment (venix12) [reply app/Application.php]: it's meant to be PM feedback

**Pull request merged**